Feat/fast faucet

meowsbits opened the pull request:
- Creates `-syncmode=[light|full|fast]` flag, with default `light` to preserve default behavior.
- Creates `-datadir` flag to allow configurable datadir placement.

meowsbits committed:
main: init --fast option for faucet

Sometimes LES is not well supported on test
networks and it can be difficult to keep a LES
instance connected.

This begins a draft of a feature to allow
the faucet to use fast/full sync which
should be a greater common denominator
on the test nets.

Date: 2020-12-17 05:56:48-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: init --syncmode.fast and --datadir flags

I'm having trouble keeping the light client
connected. I don't see why I shouldn't be able to
use a fast/full client for the faucet if I want.
This adds --syncmode.fast boolean flag to the faucet
program to support this.

Also adds a --datadir flag because I'd like to be able
to specify that.

Date: 2020-12-21 07:25:46-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: move syncmode to string flag accepting same value as geth

Allow [light|fast|full] values for a now-stringy
--syncmode flag.

This copies the flag and use from geth.

Date: 2020-12-28 07:42:30-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
Merge branch 'master' into feat/fast-faucet

Conflicts:

- [ ] 	cmd/faucet/faucet.go

meowsbits committed:
main: resolve merge conflicts

Date: 2021-01-04 07:08:56-06:00
Signed-off-by: meows <b5c6@protonmail.com>

meowsbits committed:
main: fix conflict resolution errors

Essentially typos and syntactic issues.

Date: 2021-01-04 07:15:01-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on cmd/faucet/faucet.go:
```diff
@@ -440,6 +446,13 @@ func (f *faucet) startStack(genesis *genesisT.Genesis, port int, enodes []*discv
+	} else if *syncmodeFlag == "full" {
+		cfg.SyncMode = downloader.FullSync
+		cfg.ProtocolVersions = eth.DefaultProtocolVersions
+	}
```
We might want to be explicit here that we don't handle anything else than "light".

Maybe we can add something like:
```go
} else if *syncmodeFlag != "light" {
  return fmt.Error("Supported Sync modes are one of [light | fast | full].")
}
```

meowsbits replied:
Yea, good call. See https://github.com/etclabscore/core-geth/pull/268/commits/c883a4a221a589ea62f3d2d14669e7435885a5d3, where I've enforced this in `auditFlagUse` and refactored a little to both improve legibility and remove useless `backend` construction (which is only used for `ethstats` reporting).

ziogaschr replied:
Nice refactor man. It's more readable now

ziogaschr reviewed (commented):
Please also have a look at https://github.com/etclabscore/core-geth/pull/269/files#r551442004

meowsbits committed:
main: refactor syncmode configuration switching

Using switch statements instead of ifs hopefully
yield more structured, legible code.

A useless initialization of backend values is
removed by first checking for the presence of the
statsFlag, since the backends are only built to
serve the ethstats reporting service.

Date: 2021-01-04 12:55:59-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr reviewed (approved):
LGTM
